Objects/dictobject.c

```diff
@@ -332,24 +332,29 @@ dictkeys_get_index(const PyDictKeysObject *keys, Py_ssize_t i)
     int log2size = DK_LOG_SIZE(keys);
     Py_ssize_t ix;
 
-    if (log2size < 8) {
-        const int8_t *indices = (const int8_t*)(keys->dk_indices);
-        ix = indices[i];
+    if (log2size <= 8) {
+        const uint8_t *indices = (const uint8_t*)(keys->dk_indices);
+        const uint8_t uix = indices[i] - DKIX_LOWEST_RESERVED;
```

Comment:
Maybe change `DKIX_LOWEST_RESERVED` to `DKIX_RESERVED_VALUES` (you choose the name)
such that it is a positive value. Adding a positive value makes the code easier to follow than subtracting a negative.

Reply:
I had similar considerations, but was one the fence about what is better.

So I have no strong feelings either way.  I'm happy to change it to what you prefer.

Reply:
Also, if we want to move this change into main, we would also want to change quite a few explanatory comments.

I haven't done that so far, because I wanted to get feedback on the general idea first.

Reply:
Done.